### DISCUSSION TITLE
Advice for real-time flight with gamepad, Python, and external positioning (motion capture)

### DISCUSSION THREAD

**Discussion by mbaytas** [bitcraze/discussions#571] [Q&A]
I'm currently building a real-time application with the Crazyflie, flying with a gamepad. We're using a Qualisys motion capture system and the [Qualisys Drone SDK](https://github.com/qualisys/qualisys_drone_sdk) to control the drone.

You can refer to the `setup()` function in the SDK's [QualisysCrazyflie class](https://github.com/qualisys/qualisys_drone_sdk/blob/main/qfly/crazyflie.py) for the settings we are using to initialize the drone.

Specifically, during flight, we're using the `extpos.send_extpos()` function for 3D positioning (rather than 6D with `send_extpose()`). (Because we have so far difficult to achieve stability with 6D positioning.) And we are using our `safe_position_setpoint()` function to send geofenced position setpoints, where the target position is calculated based on the drone's current position summed with gamepad input.

The drone performs OK when we set a low speed limit (`posCtlPid.xyVelMax` and `posCtlPid.zVelMax` parameters), but we'd like to fly at higher speeds, in which case the drone destabilizes and crashes easily.

What can be some tactics (e.g. parameter changes, adding a flow deck, things to double-check...) we can use to improve the stability of the drone under high speed and real time control?

PS. I am the developer of the SDK. Generally I'm looking for these suggestions to develop this particular application, but also to incorporate into the SDK to improve it.

**Comment by krichardsson** [comment]
Hi,

I usually also use 3D positioning rather than 6D, as errors in the attitude has a tendency to flip the Crazyflie easily. Another option is to increase the standard deviation when sending the attitude to the kalman filter (the `locSrv.extQuatStdDev` parameter), to make it less prominent.
Also check the quality of the position data that you push and set the standard deviation accordingly (the `locSrv.extPosStdDev`, 0.01 by default).
You can also play with how often you push position/pose data. I think around 5 Hz is enough, but a higher rate should not hurt.

> The drone performs OK when we set a low speed limit (posCtlPid.xyVelMax and posCtlPid.zVelMax parameters), but we'd like to fly at higher speeds, in which case the drone destabilizes and crashes easily.

How fast do you want to go? I think I have used the PID controller successfully up to at least around 3 m/s. It works OK when flying in straight lines but is not very good at advanced maneuvers.

> I'm currently building a real-time application with the Crazyflie, flying with a gamepad.

What kind of setpoints are you using? Position or velocity?
Are you using the high level commander?
